### ISSUE TITLE
Potential issue with LifecycleDailyCount

### ISSUE THREAD

**Issue by dj2** [cds-snc/covid-alert-app#1488]
# Summary | Résumé

This may or not be an issue, depending on if my understand is correct or not. In `DebugMetricsHelper.kt` we have `getLifecycleDailyCount`. The `DebugMetricsHelper` has a cache for `cachedLifecycleDailyCount` which is initially `null`. What is the lifetime of this class? Does it get re-created every time we wakeup in the background? Every time the app is opened? When does that cache value get reset, essentially.

If the cache is reset each background invocation then when we get into `getLifecycleDailyCount` we'll see the cache is `null` we then check the last cache time, and if it was today, we return 0, this will then set the cache to 1. If it wasn't today, we read the cache value. It seems like this would, potentially, reset the LifecycleDailyCount to 1 a lot more then we expect?

Could we remove the cache from this code to simplify things? (Are we caching because the retrieval from `sharedPreferences` is slow?)

Would something like the following work?

```kotlin
fun getLifecycleDailyCount(context: Context): Number {
  val sharedPreferences = context.getSharedPreferences(SharedPreferencesFileKey, Context.MODE_PRIVATE)
  val cachedCount = sharedPreferences.getInt(LifecycleDailyCountKey, 0)
 
  if (didDayChange(LifecycleDailyCountResetTimestampKey, context)) {
    cachedCount = cachedCount + 1
    with (sharedPreferences.edit()) {
      putInt(LifecycleDailyCountKey, cachedCount)
      apply()
    }
  }
  return cachedCount
}
```

(If the above works a similar change could also be applied to `getSuccessfulDailyBackgroundChecks` and possibly other methods which cache values.

**Comment by dj2** [comment]
I may have also just interpreted `dayDidChange` backwards ....

**Comment by craigzour** [comment]
> The DebugMetricsHelper has a cache for cachedLifecycleDailyCount which is initially null. What is the lifetime of this class?

We do use an `object` type instead of `class` for DebugMetricsHelper as it simplifies the use of this temporary debug service which is plugged in a bunch of different places in the codebase. The `cachedLifecycleDailyCount` will live until the process dies.

> Does it get re-created every time we wakeup in the background? Every time the app is opened?

Yes, as soon as the process is initialized.

> then check the last cache time, and if it was today, we return 0, this will then set the cache to 1. If it wasn't today, we read the cache value. It seems like this would, potentially, reset the LifecycleDailyCount to 1 a lot more then we expect?

The `didDayChange` function is here to let us know if we should reset the `cachedLifecycleDailyCount` which we want to set back to 0 every day. It is not an invalidate cache mechanism at all.

> Could we remove the cache from this code to simplify things? (Are we caching because the retrieval from sharedPreferences is slow?)

We could try to see if removing it has an impact. I originally implemented it to avoid querying the same information multiple times in a row over a few seconds of execution time.

**Comment by dj2** [comment]
It's probably low priority. Caches scare me as they're an extra level of complexity. If we can remove them without perf impact it's possibly useful to remove a source of variability.

**Comment by dj2** [comment]
Ok, another question. The object lives until the process dies, is the process a short lived background task to check for metrics that terminates as soon as it's complete or could it be associated to the main covid alert app? (Like, if I open the app on my phone and never close it, will the object just .... live forever?) If it lives forever we'll never update the value after the first check as we never invalidate the cache?

**Comment by craigzour** [comment]
@dj2 Great question! Not sure how it would behaves in this specific case where the app stays in the task switcher. Maybe @jeberhardt knows? 

_Also, just as a reminder, this is just a temporary feature that does not have an impact on the user and is just for us (developers) to understand what happens in the background._

**Comment by dj2** [comment]
Yup, realize it's just for development but want to make sure the numbers we're reporting make sense and help us understand whats happening. If this number (or any of the numbers in here protected by the cache variables) were off, would that throw off how we see the system working?

**Comment by craigzour** [comment]
Definitely! @smcmurtry Can you tell us how much of an impact it would have on the charts you generate?

**Comment by dj2** [comment]
Specifically, the following are cached:

* cachedLifecycleIdentifier
* cachedLifecycleDailyCount
* cachedSuccessfulDailyBackgroundChecks
* cachedActivationFlag
* cachedDeviceIdentifier  (In a TEST build config)

**Comment by smcmurtry** [comment]
> Can you tell us how much of an impact it would have on the charts you generate?

I'm not using the `lifecycleDailyCount` for the main charts but I was hoping to use it to estimate how many unique devices are starting background checks, by getting a daily sum of lifecycle ids that have a `lifecycleDailyCount` of 1.

**Comment by dj2** [comment]
I'd guess we should remove the caching before we start using the number to try to do analysis otherwise we won't have confidence that the number is accurate.

**Comment by dj2** [comment]
Are you linking events through the `lifecycleId`? If so, those are also cached and may suffer the same issue (which we could possibly check for if we see something like the same lifecycleId with the start step twice?

**Comment by craigzour** [comment]
I have been thinking about how it will work without the cached variables and we would have to add more logic somehow to know when we are in the same instance of the application in order to keep using the same lifecycleId for example. We can definitely spend time looking into it if it is a priority but it will not be as simple as what we thought. cc @timarney

**Comment by dj2** [comment]
Yea, the lifecycleId is a lot tricker as we don't have the ability to cycle it daily like the others. Do we have an entry point (or a couple) where we could setup the `lifecycleId` and then reset it to `null` again at the end? Basically, do we have a scope we can set/reset it inside?

**Comment by dj2** [comment]
The proposal (https://docs.google.com/document/d/1d8aGHDACL9hQBNb8BWDNDhBTZoaZR2lmjTr86r9NDFo/edit?ts=606c95a8#) has a few places it mentions we should reset the lifecycleId.

**Comment by craigzour** [comment]
Yes. Finding a place where we can reset the cached variables would be the solution to our problem. It does not impact just the `lifecycleId` because the `dailyCount` also needs to be cached for the same run cycle (it only gets increased every time the background check happens).

**Comment by dj2** [comment]
I'm not sure I understand. The caching of the two values aren't connected, at least from my reading? So, if we cache the lifecycleId, and it gets reused on the next run because the process was still running, won't we send a set of metrics with the same lifecycleId? Which, if we're then trying to see how far that run got, it will end in 2 different places?

**Comment by smcmurtry** [comment]
> Are you linking events through the lifecycleId? If so, those are also cached and may suffer the same issue (which we could possibly check for if we see something like the same lifecycleId with the start step twice?

Yes I am grouping all the data by lifecycleId. I have seen the same steps logged multiple times for the same lifecycleId. But only over the span of a few minutes.

**Comment by dj2** [comment]
(Or, is that now how lifecycleId is used?)

**Comment by dj2** [comment]
If we have retry logic that would be some of the duplication of steps. If we see the start step multiple times that would be multiple runs, I think?

**Comment by craigzour** [comment]
> I'm not sure I understand. The caching of the two values aren't connected, at least from my reading? So, if we cache the lifecycleId, and it gets reused on the next run because the process was still running, won't we send a set of metrics with the same lifecycleId? Which, if we're then trying to see how far that run got, it will end in 2 different places?

This is correct but I am just saying that the `lifecycleId` is not the only variable that requires a manual reset`.

> If we have retry logic that would be some of the duplication of steps. If we see the start step multiple times that would be multiple runs, I think?

There is no retry logic for sending debug metrics

**Comment by dj2** [comment]
For metrics like dailyCount, we already have the timestamp of the last change, so we can keep the `didDayChange` logic and update the value if the day is different. We just, don't want to cache that value, read it from the storage each time. The `lifecycleId` we want to change on every run, so we can't use the same timestamp trick to only update when needed.

**Comment by craigzour** [comment]
The `dailyCount` as well needs to be updated each time we run a background check. We just reset it to 0 every day but it gets incremented during the same day.

**Comment by timarney** [comment]
@dj2 anything further here?

**Comment by timarney** [comment]
@dj2 can this be closed now?
